core/services/run_queue.go

```diff
@@ -34,81 +35,176 @@ type RunQueue interface {
+	defer close(worker.chDone)
 
-	rq.workersWg.Add(1)
+	promNumberRunQueueWorkers.Inc()
+
+	var (
+		startOnce       sync.Once
+		chWorkerStarted = make(chan struct{})
+		chResume        = make(chan int)
+		chWorkComplete  = make(chan struct{})
+	)
+
+	// The worker goroutine accepts job run requests from the coordinator loop below.  It
+	// can accept further requests after work has begun.  Once its queue of work is empty,
+	// it sends a "work finished" message to that loop, which may or may not cause it to
+	// spin down, depending on whether further requests are already enqueued.
 	go func() {
-		for {
-			rq.workersMutex.Lock()
-			queueCount := rq.workers[runID]
-			if queueCount <= 0 {
-				delete(rq.workers, runID)
-				numberRunQueueWorkers.Set(float64(len(rq.workers)))
-				rq.workersMutex.Unlock()
-				break
-			}
-			rq.workers[runID] = queueCount - 1
-			rq.workersMutex.Unlock()
+		for numRunsRequested := range chResume {
+			startOnce.Do(func() { close(chWorkerStarted) })
 
-			if err := rq.runExecutor.Execute(run.ID); err != nil {
-				logger.Errorw(fmt.Sprint("Error executing run ", runID), "error", err)
+			for i := 0; i < numRunsRequested; i++ {
+				select {
+				case <-worker.chStop:
+					return
+				default:
+				}
+
+				promNumberRunsQueued.Inc()
+
+				if err := rq.runExecutor.Execute(runID); err != nil {
+					logger.Errorw(fmt.Sprint("Error executing run ", *runID), "error", err)
+				}
+			}
+			select {
+			case chWorkComplete <- struct{}{}:
+			case <-worker.chStop:
+				return
 			}
 		}
-
-		rq.workersWg.Done()
 	}()
+
+	// The coordinator loop accepts "do job run" requests from orchestrateWorkers, "work finished"
+	// signals from the worker goroutine, and "stop" messages from the Chainlink application,
+	// coordinating them such that we can avoid races when messages arrive simultaneously.
+	numRunsRequested := 0
+	for {
+		select {
+		case <-worker.chAddRun:
+			numRunsRequested++
+
+			select {
+			case chResume <- numRunsRequested:
+				numRunsRequested = 0
+			case <-chWorkerStarted:
+				// If we couldn't send the worker a work request, make sure it's not because it
+				// simply hasn't had a chance to start yet.  Avoids a race when .Stop() is called
+				// too quickly.
+			}
+
+		case <-chWorkComplete:
+			if numRunsRequested > 0 {
+				// If we've queued up more runs while the worker was working, keep it going.
+				chResume <- numRunsRequested
+				numRunsRequested = 0
+			} else {
+				// If we hit 0 runs, shut down the worker.
+				select {
+				case rq.chWorkerDone <- *runID:
+				case <-worker.chStop:
+				}
+				close(chResume)
+				return
+			}
+
+		case <-worker.chStop:
+			close(chResume)
+			return
+		}
+	}
 }
 
 // WorkerCount returns the number of workers currently processing a job run
 func (rq *runQueue) WorkerCount() int {
-	rq.workersMutex.RLock()
-	defer rq.workersMutex.RUnlock()
-
-	return len(rq.workers)
+	return int(atomic.LoadInt32(&rq.numWorkers))
```

Comment:
It maintains consistency to use `atomic.LoadInt32`, here, so I'm not suggesting to remove it, but does it actually improve the program's behavior? `numWorkers` could be updated right after the `return` regardless of the atomicity of the read, so I think it could easily still be stale.

Reply:
You're right, but I'm still inclined to do this for consistency's sake (it's sort of self-documenting in that sense).

Reply:
No, I think it's good this way, I'm just new to `atomic` semantics so I wanted to check my understanding.